src/urllib3/connectionpool.py

```diff
@@ -278,17 +279,23 @@ def _put_conn(self, conn):
         """
         try:
             self.pool.put(conn, block=False)
-            return  # Everything is dandy, done.
         except AttributeError:
             # self.pool is None.
-            pass
+            # Connection never got put back into the pool, close it.
+            conn and conn.close()
         except queue.Full:
-            # This should never happen if self.block == True
-            log.warning("Connection pool is full, discarding connection: %s", self.host)
+            # This should never happen if self.block == True and you got the conn from self._get_conn
 
-        # Connection never got put back into the pool, close it.
-        if conn:
-            conn.close()
+            # Connection never got put back into the pool, close it.
+            conn and conn.close()
+
+            if self.block:
+                raise FullPoolError(
```

Comment:
This block originally had a "This should never happen if self.block == True" comment. If this should never happen then if it *does* happen we should raise an exception.